Add tutorials for neow3j (#6)
neo-project/neo-dev-portal: tutorials/2022-02-09-neow3j-nep11/index.md

```java
package io.neow3j.examples.contractdevelopment.contracts;

import io.neow3j.devpack.ByteString;
import io.neow3j.devpack.Hash160;
import io.neow3j.devpack.Helper;
import io.neow3j.devpack.Iterator;
import io.neow3j.devpack.Map;
import io.neow3j.devpack.Runtime;
import io.neow3j.devpack.Storage;
import io.neow3j.devpack.StorageContext;
import io.neow3j.devpack.StorageMap;
import io.neow3j.devpack.StringLiteralHelper;
import io.neow3j.devpack.annotations.DisplayName;
import io.neow3j.devpack.annotations.ManifestExtra;
import io.neow3j.devpack.annotations.OnDeployment;
import io.neow3j.devpack.annotations.Safe;
import io.neow3j.devpack.annotations.SupportedStandards;
import io.neow3j.devpack.constants.FindOptions;
import io.neow3j.devpack.contracts.ContractManagement;
import io.neow3j.devpack.events.Event3Args;
import io.neow3j.devpack.events.Event4Args;
```

### Contract-specific Information

Annotations on top of the contract's class represent contract-specific information. The following annotations are used in the example contract:

_`@DisplayName`_

> Specifies the contract's name. If this annotation is not present, the class name is used for the contrat's name.

Reviewer: ```suggestion
It specifies the contract's name. If this annotation is not present, the class name is used for the contract's name.
```

_`@ManifestExtra`_

Adds the provided key-value pair information in the manifest's `extra` field. You can also use `@ManifestsExtras` to gather multiple `@ManifestExtra` annotations (results in the same as when using single `@ManifestExtra` annotations).

_`@SupportedStandards`_

Sets the `supportedStandards` field in the manifest.

```java
@DisplayName("FurryFriends")
@ManifestExtra(key = "author", value = "AxLabs")
@SupportedStandards("NEP-11")
public class NonFungibleToken {
```

### Constants

You can set a constant value for the contract by using `final` variables. These values are always loaded when the contract is called and cannot be changed once the contract is deployed.

> > Note: The contract owner of this example contract is fixed (i.e., it is a `final` variable). If you intend to provide a functionality to change such a variable, you should not store it as a `final` variable. Rather, you would store it as a value in the storage, that you then can change through a method.

Reviewer: ```suggestion
> Note: The contract owner of this example contract is fixed (i.e., it is a `final` variable). If you intend to provide a way to change such a variable, you should not store it as a `final` variable. Rather, you would store it as a value in the storage, that can be changed through a method.
```

-

```java
    static final Hash160 contractOwner = StringLiteralHelper.addressToScriptHash("NM7Aky765FG8NhhwtxjXRx7jEL1cnw7PBP");

    static final StorageContext ctx = Storage.getStorageContext();
    static final StorageMap contractMap = new StorageMap(ctx, 0);
    static final StorageMap registryMap = new StorageMap(ctx, 1);
    static final StorageMap ownerOfMap = new StorageMap(ctx, 2);
    static final StorageMap balanceMap = new StorageMap(ctx, 3);

    // Keys of key-value pairs in NFT properties
    static final String propName = "name";
    static final String propDescription = "description";
    static final String propImage = "image";
    static final String propTokenURI = "tokenURI";

    static final StorageMap propertiesNameMap = new StorageMap(ctx, 8);
    static final StorageMap propertiesDescriptionMap = new StorageMap(ctx, 9);
    static final StorageMap propertiesImageMap = new StorageMap(ctx, 10);
    static final StorageMap propertiesTokenURIMap = new StorageMap(ctx, 11);

    static final byte[] totalSupplyKey = new byte[]{0x10};
    static final byte[] tokensOfKey = new byte[]{0x11};
```

### NEP-11 Methods

> The required NEP-11 methods are implemented in the following. If a method does not change the state of the contract (i.e., it is just used for reading), it can be annotated with the `@Safe` annotation. Out of the required NEP-11 methods, only the `transfer` method should be writing to the contract and is thus not annotated as safe.

Reviewer: ```suggestion
The required NEP-11 methods are implemented in the following. If a method does not change the state of the contract (i.e., it is just used for reading), it can be annotated with the `@Safe` annotation. Out of the required NEP-11 methods, only the `transfer` method should be writing to the contract's storage and, thus, is not annotated as safe.
```

> > Note: Java's `assert` statement is used in neow3j to throw exceptions on the NeoVM. Thus, there is no need of using an `if` statement with a followed `throw new Exception()` statement. Although, it may still be used like that. It is up to the developer's preference, as it results in the exact same NeoVM code once compiled.

Reviewer: ```suggestion
> Note: Java's `assert` statement is used in neow3j to throw exceptions on the NeoVM. Thus, there is no need of using an `if` statement with a followed `throw new Exception()` statement. Although, it may still be used like that. It is up to the developer to go either way, as it results in the exact same NeoVM code once compiled.
```

    }

    @Safe
    public static int balanceOf(Hash160 owner) {
        return balanceMap.getIntOrZero(owner.toByteArray());
    }

    @Safe
    public static Iterator<ByteString> tokensOf(Hash160 owner) {
        return (Iterator<ByteString>) Storage.find(
                ctx.asReadOnly(), createTokensOfPrefix(owner), FindOptions.RemovePrefix);
    }

    public static boolean transfer(Hash160 to, ByteString tokenId, Object data) {
        Hash160 owner = ownerOf(tokenId);
        assert owner != null : "This token id does not exist.";
        throwIfSignerIsNotOwner(owner);

        ownerOfMap.put(tokenId, to.toByteArray());

        new StorageMap(ctx, createTokensOfPrefix(owner)).delete(tokenId);
        new StorageMap(ctx, createTokensOfPrefix(to)).put(tokenId, 1);

        decreaseBalanceByOne(owner);
        increaseBalanceByOne(to);

        onTransfer.fire(owner, to, 1, tokenId);
        return true;
    }

    @Safe
    public static Hash160 ownerOf(ByteString tokenId) {
        ByteString owner = ownerOfMap.get(tokenId);
        if (owner == null) {
            return null;
        }
        return new Hash160(owner);
    }
```

### NEP-11 Optional Methods

> The NEP-11 standard states two optional methods (i.e., if methods with these names and parameters is implemented, it must follow the standard.). Following are the methods `tokens` and `properties` of the example contract shown. The `tokens` method iterates through the `registryMap` and returns an `Iterator` based on the key-value pairs that are found in the registry. The `properties` method returns a map of the provided token's properties stored in the contract's storage. This includes its name, and if present its description, image, and URI.

Reviewer: ```suggestion
The NEP-11 standard describes two optional methods called `tokens()` and `properties()` -- meaning that if methods with these names and parameters are implemented, they must follow the standard. Below you can see the implementation of these two methods. The `tokens` method iterates through the `registryMap` and returns an `Iterator` based on the key-value pairs that are found in the registry. The `properties` method returns a map of the provided token's properties stored in the contract's storage. This includes its name, and if present its description, image, and URI.
```

```java
    @Safe
    public static Iterator<Iterator.Struct<ByteString, ByteString>> tokens() {
        return (Iterator<Iterator.Struct<ByteString, ByteString>>) registryMap.find(FindOptions.RemovePrefix);
    }

    @Safe
    public static Map<String, String> properties(ByteString tokenId) {
        Map<String, String> p = new Map<>();
        ByteString tokenName = propertiesNameMap.get(tokenId);
        assert tokenName != null : "This token id does not exist.";

        p.put(propName, tokenName.toString());
        ByteString tokenDescription = propertiesDescriptionMap.get(tokenId);
        if (tokenDescription != null) {
            p.put(propDescription, tokenDescription.toString());
        }
        ByteString tokenImage = propertiesImageMap.get(tokenId);
        if (tokenImage != null) {
            p.put(propImage, tokenImage.toString());
        }
        ByteString tokenURI = propertiesTokenURIMap.get(tokenId);
        if (tokenURI != null) {
            p.put(propTokenURI, tokenURI.toString());
        }
        return p;
    }
```

> ### Events

Reviewer: ```suggestion
### Events

```

The NEP-11 standard requires an event `Transfer` that contains the values `from`, `to`, `amount`, and `tokenId`. For this, the class `Event4Args` can be used with the annotation `@DisplayName` to set the event's name that will be shown in the manifest and notifications when it has been fired. The event `Mint` is an custom event that is fired whenever a new NFT is minted.

```java
    @DisplayName("Transfer")
    static Event4Args<Hash160, Hash160, Integer, ByteString> onTransfer;

    @DisplayName("Mint")
    private static Event3Args<Hash160, ByteString, Map<String, String>> onMint;
```

> When an event should be fired, its `fire` method can be called with the corresponding arguments. For example, the `Transfer` event should be fired whenever a transfer happens.

Reviewer: ```suggestion
An event variable can effectively fire an event by using the `fire()` method with the corresponding arguments. For example, the `Transfer` event (represented by the `onTransfer` variable) should be fired whenever a transfer happens.
```

```java
    onTransfer.fire(owner, to, 1, tokenId);
```

### Deploy

> Within the deployment transaction of the contract, the contract data is first stored on the blockchain and then the native contract `ContractManagement` calls the contract's `deploy` method. In neow3j, that method is marked with the annotation `@OnDeployment`. In the example, when the contract is deployed, the total supply is set to 0.

Reviewer: ```suggestion
Once a deployment transaction is made (containing the contract and other params), the contract data is first stored on the blockchain and then the native contract `ContractManagement` calls the smart contract's `deploy` method. In neow3j, that method is marked with the annotation `@OnDeployment`. In the example, when the contract is deployed, the total supply is set to 0.
```

```java
    @OnDeployment
    public static void deploy(Object data, boolean update) {
        if (!update) {
            throwIfSignerIsNotContractOwner();
            contractMap.put(totalSupplyKey, 0);
        }
    }
```

### Update and Destroy

> In order to update or destroy the contract, the following two simple methods first check that the contract owner witnessed the transaction and then the appropriate method is called on the native contract `ContractManagement`.

Reviewer: maybe we should re-phrase this paragrah.

        }
        if (properties.containsKey(propTokenURI)) {
            String tokenURI = properties.get(propTokenURI);
            propertiesTokenURIMap.put(tokenId, tokenURI);
        }

        registryMap.put(tokenId, tokenId);
        ownerOfMap.put(tokenId, owner.toByteArray());
        new StorageMap(ctx, createTokensOfPrefix(owner)).put(tokenId, 1);

        increaseBalanceByOne(owner);
        incrementTotalSupplyByOne();
        onMint.fire(owner, tokenId, properties);
    }

    public static boolean burn(ByteString tokenId) {
        Hash160 owner = ownerOf(tokenId);
        assert owner != null : "This token id does not exist.";
        throwIfSignerIsNotOwner(owner);

        registryMap.delete(tokenId);
        propertiesNameMap.delete(tokenId);
        propertiesDescriptionMap.delete(tokenId);
        propertiesImageMap.delete(tokenId);
        propertiesTokenURIMap.delete(tokenId);
        ownerOfMap.delete(tokenId);

        new StorageMap(ctx, createTokensOfPrefix(owner)).delete(tokenId);
        decreaseBalanceByOne(owner);
        decrementTotalSupplyByOne();
        return true;
    }
```

### Private Helper Methods

> Private methods can be used to extract code that is used internally in the contract. The following private methods are used in the NEP-17 example contract. For example, in order to prevent writing the same exception message to check the witness of the owner, the private method `throwIfSignerIsNotContractOwner` allows to write this code only once.

Reviewer: ```suggestion
Private methods can be used to, for example, simplify and make the smart contract a bit more readable. The following private methods are used in the NEP-17 example contract. For example, in order to prevent writing the same exception message to check the witness of the owner, the private method `throwIfSignerIsNotContractOwner` allows writing this code only once.
```

```java
    private static void throwIfSignerIsNotContractOwner() {
        assert Runtime.checkWitness(contractOwner) : "No authorization.";
    }

    private static void throwIfSignerIsNotOwner(Hash160 owner) {
        assert Runtime.checkWitness(owner) : "No authorization.";
    }

    private static void increaseBalanceByOne(Hash160 owner) {
        balanceMap.put(owner.toByteArray(), balanceOf(owner) + 1);
    }

    private static void decreaseBalanceByOne(Hash160 owner) {
        balanceMap.put(owner.toByteArray(), balanceOf(owner) - 1);
    }

    private static void incrementTotalSupplyByOne() {
        int updatedTotalSupply = contractMap.getInt(totalSupplyKey) + 1;
        contractMap.put(totalSupplyKey, updatedTotalSupply);
    }

    private static void decrementTotalSupplyByOne() {
        int updatedTotalSupply = contractMap.getInt(totalSupplyKey) - 1;
        contractMap.put(totalSupplyKey, updatedTotalSupply);
    }

    private static byte[] createTokensOfPrefix(Hash160 owner) {
        return Helper.concat(tokensOfKey, owner.toByteArray());
    }
```

## 5. Compile the Contract

> The contract can be compiled using the gradle plugin. First, set the `className` in the file `gradle.build` to the contract's class name. Then, the gradle task `neow3jCompile` can be executed from the project's root path to compile the contract.

Reviewer: Can we add references to other articles? For example, "As seen in this article, the smart contract is compiled using the Gradle plugin." -- adding the appropriate reference link.

Reviewer 2: I prefer not to add additional links (referring to claude's comment below ;)). Imo, the steps for compiling a contract are explained well enough here.

Reviewer: alright. See: https://github.com/neo-project/neo-dev-portal/pull/6#issuecomment-1036035393
